### PR TITLE
refactor(neighbors): centralize JAX registration

### PR DESCRIPTION
**Pull request by zubatyuk**
# ALCHEMI Toolkit-Ops Pull Request

## Description

Consolidate JAX neighbor-list registration behind private, schema-aware lazy registries.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Performance improvement
- [ ] Documentation update
- [x] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

None.

## Changes Made

- Add private immutable registration specifications and schema-aware lazy wrapper construction for naive, cell-list, and cluster-tile JAX neighbor bindings.
- Preserve existing direct-kernel and callback execution paths, output ABI ordering, JAX gradient reconstruction, and cluster-tile graph preload behavior.
- Add registration, cache, preload, graph-capture, and wrapper-isolation regression coverage.

## Testing

- [x] Unit tests pass locally (`make pytest`) — Not run.
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations.

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

Cold-import benchmarks for the eight JAX neighbor modules improved by ~45% versus the clean baseline. 

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

**Review by laserkelvin — COMMENTED** [review]
I found one first-use registration issue and two gaps in the tests that protect the PR's lazy-registration and graph-capture contracts; details inline.

File: nvalchemiops/jax/neighbors/naive.py [hunk 1]
```diff
@@ -2423,48 +2103,28 @@ def naive_neighbor_list(
...
-            _jax_fill_pbc = _jax_fill_naive_pbc_f32
-            _jax_fill_pbc_prewrapped = _jax_fill_naive_pbc_prewrapped_f32
-            _jax_fill_selective = _jax_fill_naive_selective_f32
-            _jax_fill_pbc_selective = _jax_fill_naive_pbc_selective_f32
-            _jax_fill_pbc_prewrapped_selective = (
-                _jax_fill_naive_pbc_prewrapped_selective_f32
-            )
         _jax_wrap_single = _jax_wrap_positions_single_f32
         positions = positions.astype(jnp.float32)
+    _jax_fill = _DIRECT_NAIVE_KERNELS[("none", False, bool(half_fill))][positions.dtype]
```
Comment (laserkelvin) [inline nvalchemiops/jax/neighbors/naive.py]: `issue:` These lookups materialize all six direct registrations before the code selects direct, tile, or WARP-graph dispatch below; the same pattern recurs in the batch and dual-cutoff bindings. I assume lazy registration should pay only for the specialization a call uses. Move the lookups into the chosen direct branch and resolve only the selected PBC/selective variant.

File: test/neighbors/bindings/jax/test_cluster_tile.py [hunk 2]
```diff
@@ -175,6 +176,37 @@ def test_non_aligned_N_accepts_padding(self):
         assert bool(jnp.all(nn >= 0))
 
 
+class TestClusterTileGraphPreload:
+    """Exercise the public WARP graph callback after its kernel preload."""
+
+    def test_jitted_matrix_query_preloads_and_executes_warp_graph_callback(self):
```
Comment (laserkelvin) [inline test/neighbors/bindings/jax/test_cluster_tile.py:209]: `suggestion:` This JIT test covers only the unbatched, non-selective matrix preload path, so the distinct batched-matrix and COO preload branches can regress while this stays green. I assume the graph-capture guarantee applies to each registered cluster-tile callback shape. Add JIT cases for unbatched COO, batched matrix, and batched segmented COO.

File: test/neighbors/bindings/jax/test_registration.py [hunk 3]
```diff
@@ -0,0 +1,1542 @@
...
+
+class TestNaiveBindingRegistrations:
+    """Test lazy registry wiring in the naive JAX bindings."""
+
+    @staticmethod
+    def _import_binding(module_name: str):
+        """Import one naive JAX binding module."""
+        return importlib.import_module(f"nvalchemiops.jax.neighbors.{module_name}")
+
+    @pytest.mark.parametrize(
```
Comment (laserkelvin) [inline test/neighbors/bindings/jax/test_registration.py]: `suggestion:` This import-laziness matrix omits `naive`, and the cell-list import test below spies only on the build getter, leaving query registries unprotected. I assume import-time deferral is a contract for every registry family introduced here. Extend the parametrization to all four naive modules and both cell-list build/query getters, ideally with fresh imports.

**Comment by laserkelvin** [comment]
My main issue right now is that the PR introduces a lot of intermediate, private data classes that make it difficult to maintain. I suggest trying to consolidate the caller-facing model to two concepts:

1. A lazy dtype-specific kernel registration.
2. A graph callback bundled with its preload operation.

Here's a mapping between existing structures, and something we could try and collapse them with:

| Current structure | Recommendation |
| --- | --- |
| `_JaxOutputSchema` | Remove; an output tuple already provides `len(outputs)` |
| `_NaiveJaxKernelSpec` | Hide behind a `lazy_naive_kernel(...)` factory |
| `_CellListBuildJaxSpec` | Replace with a table-driven build-stage factory |
| `_CellListQueryJaxSpec` | Hide behind a `lazy_cell_query(...)` factory |
| `_ClusterTileBuildJaxSpec` | Bundle registration and preload behavior |
| `_ClusterTileQueryJaxSpec` | Bundle registration and preload behavior |
| `_LazyDtypeRegistrations` | Keep as the core abstraction, renamed `_LazyJaxKernel` |

The core object could look roughly like:

~~~python
Outputs = tuple[str, ...]


class _LazyJaxKernel:
    def __init__(
        self,
        build: Callable[[type], tuple[wp.Kernel, Outputs]],
        dtype_map: Mapping[object, type],
    ) -> None:
        self._build = build
        self._dtype_map = normalize_dtype_map(dtype_map)
        self._cache = {}

    def __getitem__(self, jax_dtype):
        wp_dtype = self._dtype_map[jnp.dtype(jax_dtype)]
        if wp_dtype not in self._cache:
            kernel, outputs = self._build(wp_dtype)
            self._cache[wp_dtype] = jax_kernel(
                kernel,
                num_outputs=len(outputs),
                in_out_argnames=list(outputs),
                enable_backward=False,
            )
        return self._cache[wp_dtype]
~~~

Family factories then contain their own validation and output policy:

~~~python
def lazy_naive_kernel(
    *,
    operation: NaiveOperation,
    batched: bool,
    pbc_mode: PbcMode,
    selective: bool = False,
    half_fill: bool = False,
    geometry: bool = False,
) -> _LazyJaxKernel:
    validate_naive_options(...)

    outputs = naive_outputs(
        operation=operation,
        pbc_mode=pbc_mode,
        geometry=geometry,
    )

    def build(wp_dtype):
        kernel = get_naive_kernel(
            wp_dtype,
            operation=operation,
            batched=batched,
            pbc_mode=pbc_mode,
            selective=selective,
            half_fill=half_fill,
            return_vectors=geometry,
            return_distances=geometry,
        )
        return kernel, outputs

    return _LazyJaxKernel(build, _DTYPE_MAP)
~~~

...and callers no longer need to import or construct `_NaiveJaxKernelSpec`:

~~~python
_DIRECT_NAIVE_KERNELS = {
    (pbc_mode, selective, half_fill): lazy_naive_kernel(
        operation="single_cutoff",
        batched=False,
        pbc_mode=pbc_mode,
        selective=selective,
        half_fill=half_fill,
    )
    for pbc_mode in PBC_MODES
    for selective in (False, True)
    for half_fill in (False, True)
}
~~~

The most difficult code is the parallel state in `cluster_tile.py`: seven query specifications are followed by seven corresponding registered callables, while dispatch later selects a spec separately from its callable and preloader. We could potentially replace those parallel objects with one cohesive result:

~~~python
@dataclass(frozen=True)
class _GraphRegistration:
    callable: Any
    preload: Callable[[], None]
~~~

Then expose semantic factory functions:

~~~python
_CLUSTER_QUERIES = {
    "matrix": cluster_matrix_registration(
        _query_cluster_tile_callback,
        batched=False,
    ),
    "selective": cluster_matrix_registration(
        _query_cluster_tile_selective_callback,
        batched=False,
        selective=True,
    ),
    "dual": cluster_matrix_registration(
        _query_cluster_tile_dual_callback,
        batched=False,
        dual_cutoff=True,
    ),
    "coo": cluster_coo_registration(
        _query_cluster_tile_coo_callback,
        batched=False,
    ),
}
~~~

Dispatch becomes:

~~~python
registration = _CLUSTER_QUERIES[query_mode]
registration.preload()
return registration.callable(...)
~~~

That prevents a specification, callable, and preload call from drifting apart. It would also naturally help fix the review finding where all lazy direct variants are currently resolved before the actual execution path is selected.

**Comment by laserkelvin** [comment]
/ok to test [d4b35f1](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/130/commits/d4b35f1f0d3978a9307da42432bf2f538d3fe5e1)

File: nvalchemiops/jax/neighbors/naive_dual_cutoff.py [hunk 4]
```diff
@@ -32,185 +32,31 @@
...
-) = build_naive_kernel_tables(
-    "dual_cutoff", batched=False, dtypes=_DTYPE_TO_NAIVE_DUAL_KERNELS
-)
-
-__all__ = ["naive_neighbor_list_dual_cutoff"]
 
-# ==============================================================================
-# JAX Kernel Wrappers
-# ==============================================================================
+def _direct_dual_registrations(*, pbc_mode: str, selective: bool = False):
```
Comment (laserkelvin) [inline nvalchemiops/jax/neighbors/naive_dual_cutoff.py]: This is mostly style, but I rather we don't need this function; just run the loop a little more verbosely

Comment (zubatyuk) [reply nvalchemiops/jax/neighbors/naive_dual_cutoff.py]: Fixed in 12ff27cb1215fffdd337ed861997f3c49db21380.

**Comment by laserkelvin** [comment]
/ok to test [5d9ed86](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/130/commits/5d9ed869b058270ed65f3708e1e120ba39573c47)

**Review by laserkelvin — APPROVED** [review]

**Comment by laserkelvin** [comment]
Resolved changelog conflicts; bypassing re-test
